sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1889,13 +1889,27 @@ class Analyzer(override val catalogManager: CatalogManager)
   private def resolveExpression(
       expr: Expression,
       resolveColumnByName: Seq[String] => Option[Expression],
-      resolveColumnByOrdinal: Int => Attribute,
+      resolveByOrdinalCandidates: () => Seq[Attribute],
```

Comment:
why don't just take `Seq[Attribute]`? Now we don't need ordinal param, and the candidates are static.

Reply:
BTW as it doesn't rely on ordinal now,  maybe just `attrCandidates` instead of `resolveByOrdinalCandidates`.

Reply:
This needs to be lazy because of `assert(q.children.length == 1)` in https://github.com/apache/spark/pull/31930/files#diff-ed19f376a63eba52eea59ca71f3355d4495fad4fad4db9a3324aade0d4986a47R1988

We shouldn't fail if we don't need to resolve by ordinal.